src/test/java/ir/sahab/kafka/reader/KafkaProtoParquetWriterTest.java

```diff
@@ -128,47 +118,60 @@ public void testMaxOpenDuration() throws Throwable {
-            // dateTimePattern is not given, checking if file is created at root of target directory
-            assertEquals(directory.getPath(), file.getPath().getParent());
-        }
-        List<SampleMessage> readRecords =
-                ParquetTestUtils.readParquetFiles(hdfsConfig, files, SampleMessage.class);
+        assertThat(files.get(0).getLen(), greaterThan(0L));
+        // dateTimePattern is not given, checking if file is created at root of target directory
+        assertEquals(directory.getPath(), files.get(0).getPath().getParent());
+
+        List<SampleMessage> readRecords = ParquetTestUtils.readParquetFiles(hdfsConfig, files, SampleMessage.class);
         assertEquals(messages.size(), readRecords.size());
         assertThat(messages, containsInAnyOrder(readRecords.toArray(new SampleMessage[0])));
     }
 
     @Test
     public void testMaxFileSize() throws Throwable {
-        final int maxFileSize = 100 * 1024;
-        final int messageCount = 1000;
-        Builder<SampleMessage> builder =
-                new Builder<SampleMessage>()
-                        .hadoopConf(hdfsConfig)
-                        .instanceName(INSTANCE_NAME)
-                        .topicName(TOPIC)
-                        .consumerConfig(consumerConfig)
-                        .maxFileOpenDurationSeconds(DEFAULT_MAX_FILE_OPEN_DURATION_SECONDS)
-                        .targetDir(targetPath)
-                        .protoClass(SampleMessage.class)
-                        .parser(SampleMessage.PARSER)
-                        .blockSize(10 * 1024)
-                        .maxFileSize(maxFileSize);
+        final int blockSize = 10 * 1024;
+        final int maxFileSize = 10 * blockSize;
+        final int messageCount = 1_000;
+        final int numberOfFilesWithMaxFileSize = 2;
+
+        int numberOfMessagesSent = 0;
+
+        Builder<SampleMessage> builder = createCommonBuilder()
+                .blockSize(blockSize)
+                .maxFileSize(maxFileSize);
+
         try (KafkaProtoParquetWriter<SampleMessage> writer = builder.build()) {
             writer.start();
-            while (findFiles(DEFAULT_PARQUET_FILE_EXTENSION).size() < 2) {
-                sendSampleMessages(messageCount);
+
+            while (findFiles(DEFAULT_PARQUET_FILE_EXTENSION).size() < numberOfFilesWithMaxFileSize) {
+                if (writer.getTotalWrittenRecords() != numberOfMessagesSent) {
+                    Thread.sleep(100L);
+                } else {
+                    sendSampleMessages(messageCount);
+                    numberOfMessagesSent += messageCount;
+                }
             }
+
+            // Make sure all records are written.
+            while (writer.getTotalWrittenRecords() != numberOfMessagesSent) {
+                Thread.sleep(100L);
+            }
+
+            // Wait for the last file to be written.
+            waitForFiles(numberOfFilesWithMaxFileSize + 1, DEFAULT_PARQUET_FILE_EXTENSION);
```

Comment:
```suggestion
           ```suggestion
           // Make sure that Parquet Writer has written the expected files. 
           int numMessages = 1000;
           int totalMessagesSent = 0;
             while (findFiles(DEFAULT_PARQUET_FILE_EXTENSION).size() < numberOfFilesWithMaxFileSize) {
                sendSampleMessages(numMessages);
                totalMessagesSent += numMessages;
            }
            
            // If there are some records remained, wait for them to be written too.
            boolean extraFileExists;
            if (writer.getTotalWrittenRecords() != totalMessagesSent) {
                    extraFileExists = true;
                    waitForFiles(numberOfFilesWithMaxFileSize + 1, DEFAULT_PARQUET_FILE_EXTENSION);
            }       
```
I changed it a little bit to make the steps and the reasons behind them, more clear.

Reply:
Done